src/main/netplay.c

```diff
@@ -0,0 +1,602 @@
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.          *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#define M64P_CORE_PROTOTYPES 1
+#include "api/callbacks.h"
+#include "main.h"
+#include "util.h"
+#include "plugin/plugin.h"
+#include "backends/plugins_compat/plugins_compat.h"
+#include "netplay.h"
+
+#include <SDL_net.h>
+
+static int l_canFF;
+static int l_netplay_controller;
+static int l_netplay_control[4];
+static UDPsocket l_udpSocket;
+static TCPsocket l_tcpSocket;
+static int l_udpChannel;
+static int l_spectator;
+static int l_netplay_is_init = 0;
+static uint32_t l_vi_counter;
+static uint8_t l_status;
+static uint32_t l_reg_id;
+static struct controller_input_compat *l_cin_compats;
+static uint8_t l_plugin[4];
+static uint8_t l_buffer_target;
+
+m64p_error netplay_start(const char* host, int port)
+{
+    if (SDLNet_Init() < 0)
+    {
+        DebugMessage(M64MSG_ERROR, "Netplay: Could not initialize SDL Net library");
+        return M64ERR_SYSTEM_FAIL;
+    }
+
+    l_udpSocket = SDLNet_UDP_Open(0);
+    if (l_udpSocket == NULL)
+    {
+        DebugMessage(M64MSG_ERROR, "Netplay: UDP socket creation failed");
+        return M64ERR_SYSTEM_FAIL;
+    }
+
+    IPaddress dest;
+    SDLNet_ResolveHost(&dest, host, port);
+
+    l_udpChannel = SDLNet_UDP_Bind(l_udpSocket, -1, &dest);
+    if (l_udpChannel < 0)
+    {
+        DebugMessage(M64MSG_ERROR, "Netplay: could not bind to UDP socket");
+        SDLNet_UDP_Close(l_udpSocket);
+        l_udpSocket = NULL;
+        return M64ERR_SYSTEM_FAIL;
+    }
+
+    l_tcpSocket = SDLNet_TCP_Open(&dest);
+    if (l_tcpSocket == NULL)
+    {
+        DebugMessage(M64MSG_ERROR, "Netplay: could not bind to TCP socket");
+        SDLNet_UDP_Close(l_udpSocket);
+        l_udpSocket = NULL;
+        return M64ERR_SYSTEM_FAIL;
+    }
+
+    for (int i = 0; i < 4; ++i)
+    {
+        l_netplay_control[i] = -1;
+        l_plugin[i] = 0;
+    }
+
+    l_canFF = 0;
+    l_netplay_controller = 0;
+    l_netplay_is_init = 1;
+    l_spectator = 1;
+    l_vi_counter = 0;
+    l_status = 0;
+    l_reg_id = 0;
+
+    return M64ERR_SUCCESS;
+}
+
+m64p_error netplay_stop()
+{
+    if (l_udpSocket == NULL)
+        return M64ERR_INVALID_STATE;
+    else
+    {
+        char output_data[5];
+        output_data[0] = 7; //Notify server that we are disconnecting
```

Comment:
I still think we should use named constants instead of magic numbers. At least for packet type.